src/base/filesystemwatcher.cpp

```diff
@@ -29,159 +30,522 @@
         {
-            QFile::rename(torrentPath, torrentPath + ".qbt_rejected");
+            if (!QFile::exists(torrentPath))
+                return true;
+
+            if (BitTorrent::TorrentInfo::loadFromFile(torrentPath).isValid())
+            {
+                BitTorrent::AddTorrentParams addTorrentParams = options.addTorrentParams;
+                if (options.recursive)
+                {
+                    const QString subdirPath = dir.relativeFilePath(QFileInfo(torrentPath).canonicalPath());
+                    addTorrentParams.savePath = QDir(addTorrentParams.savePath).filePath(subdirPath);
+                }
+
+                addTorrentToSession(torrentPath, addTorrentParams);
+                return true;
+            }
+
+            if (value >= MAX_PARTIAL_RETRIES)
+            {
+                QFile::rename(torrentPath, torrentPath + ".qbt_rejected");
+                return true;
+            }
+
+            ++value;
+            return false;
+        });
+
+        if (partialTorrents.isEmpty())
             return true;
-        }
 
-        ++value;
         return false;
     });
 
     // Stop the partial timer if necessary
     if (m_partialTorrents.empty())
+        m_partialTorrentTimer->stop();
+    else
+        m_partialTorrentTimer->start(WATCH_INTERVAL);
+}
+
+void FileSystemWatcher::addTorrentToSession(const QString &filePath, const BitTorrent::AddTorrentParams &params)
+{
+    if (filePath.endsWith(QLatin1String(".magnet"), Qt::CaseInsensitive))
     {
-        m_partialTorrentTimer.stop();
-        qDebug("No longer any partial torrent.");
+        QFile file {filePath};
+        if (file.open(QIODevice::ReadOnly | QIODevice::Text))
+        {
+            QTextStream str {&file};
+            while (!str.atEnd())
+                BitTorrent::Session::instance()->addTorrent(str.readLine(), params);
+
+            file.close();
+            Utils::Fs::forceRemove(filePath);
+        }
+        else
+        {
+            LogMsg(tr("Failed to open magnet file: %1").arg(file.errorString()));
+        }
     }
     else
     {
-        qDebug("Still %d partial torrents after delayed processing.", m_partialTorrents.count());
-        m_partialTorrentTimer.start(WATCH_INTERVAL);
+        const auto torrentInfo = BitTorrent::TorrentInfo::loadFromFile(filePath);
+        if (torrentInfo.isValid())
+        {
+            BitTorrent::Session::instance()->addTorrent(torrentInfo, params);
+            Utils::Fs::forceRemove(filePath);
+        }
+        else
+        {
+            LogMsg(tr("Ignoring incomplete torrent file: %1").arg(filePath));
+        }
     }
+}
 
-    // Notify of new torrents
-    if (!noLongerPartial.isEmpty())
-        emit torrentsAdded(noLongerPartial);
+void FileSystemWatcher::addWatchedFolder(const QString &path, const FileSystemWatcher::WatchedFolderOptions &options)
+{
+#if !defined Q_OS_HAIKU
```

Comment:
I'm not sure we really care about Haiku... Maybe I should just delete this condition?